dev/buildtool/spin_commands.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2018 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implements SpinCLI support commands for buildtool."""
+
+import logging, os
+
+from collections import namedtuple
+
+from buildtool import (
+    BomSourceCodeManager,
+    RepositoryCommandFactory,
+    RepositoryCommandProcessor,
+
+    check_subprocess)
+
+from google.cloud import storage
+
+
+class DistArch(namedtuple('DistArch', ['dist', 'arch'])):
+  """Describes an distribution/architecture pair.
+  """
+  def __repl__(self):
+    return '{},{}'.format(self.dist, self.arch)
+
+
+DIST_ARCH_LIST = [
+  DistArch('darwin', 'amd64'),
+  DistArch('linux', 'amd64')
+]
+
+
+class BuildSpinCommand(RepositoryCommandProcessor):
+  def __init__(self, factory, options, **kwargs):
+    super(BuildSpinCommand, self).__init__(factory, options, **kwargs)
+    self.__gcs_uploader = SpinGcsUploader(options)
+
+  def _do_can_skip_repository(self, repository):
+    return repository.name != 'spin'
+
+  def _do_repository(self, repository):
+    """Implements RepositoryCommandProcessor interface."""
+    name = repository.name
+    build_version = self.scm.get_repository_service_build_version(repository)
+
+    if name == 'spin':
```

Comment:
If you pass source_repository_names=['spin'] to the command constructor (or factory constructor to pass here) then the command will filter for you so this will only be called for 'spin'.